kevm-pyk/src/tests/integration/test_prove.py

```diff
@@ -118,31 +119,58 @@ class Target(NamedTuple):
     main_file: Path
     main_module_name: str
 
+    @property
+    def name(self) -> str:
+        """
+        Target's name is the two trailing path segments and the main module name
+        """
+        return f'{self.main_file.parts[-2]}-{self.main_file.stem}-{self.main_module_name}'
+
     def __call__(self, output_dir: Path) -> Path:
-        return kevm_kompile(
-            output_dir=output_dir / 'kompiled',
-            target=KompileTarget.HASKELL,
-            main_file=self.main_file,
-            main_module=self.main_module_name,
-            syntax_module=self.main_module_name,
-            debug=True,
-        )
+        with FileLock(str(output_dir) + '.lock'):
+            return kevm_kompile(
+                output_dir=output_dir / 'kompiled',
+                target=KompileTarget.HASKELL,
+                main_file=self.main_file,
+                main_module=self.main_module_name,
+                syntax_module=self.main_module_name,
+                debug=True,
+            )
+
+
+@pytest.fixture(scope='module')
+def kompiled_target_cache(kompiled_targets_dir: Path) -> tuple[Path, dict[str, Path]]:
+    """
+    Populate the cache of kompiled definitions from an existing file system directory. If the cache is hot, the `kompiled_target_for` fixture will not containt a call to `kompile`, saving an expesive call to the K frontend.
+    """
+    cache_dir = kompiled_targets_dir / 'target'
+    cache: dict[str, Path] = {}
+    if cache_dir.exists():  # cache dir exists, populate cache
+        for file in cache_dir.iterdir():
+            if file.is_dir():
+                # the cache key is the name of the target, which is the filename by-construction.
+                cache[file.stem] = file / 'kompiled'
```

Comment:
Initializing a cache might be unnecessary, you can just do the directory lookup when you request the kompiled directory.

Reply:
I don't understand what you mean: could you elaborate?

 My goals are:
* keep using `cache` in so that the old behavior of the `kompiled_target_for` fixture is retained,
* make sure that `cache` is never actually updated when `--kompiled-targets-dir` is passed.

Reply:
As I see it, tests basically require an interface that takes a `Target` and returns a `Path` to the kompiled target (in the actual tests there might be another indirection from spec files to targets, but I'll not consider this for simplicity).

So you just have to ensure that `Target -> Path` is reproducible between runs, then accommodating both use cases (prekompiled vs on-the-fly) is straightforward.

* If the `Path` does not exist, lock and kompile
* Otherwise, return the `Path`.

For prekompiled mode, you pass in a non-empty directory as base. For on-the-file mode, you pass in an empty temporary directory.

`cache`, etc. can be dropped, the actual cache is the file system.

Reply:
Yeah, that sound good, I understand what you mean. However, I'd rather get my incremental change merged quickly and now spend too much time refactoring the test suite. Would that be OK if I let you refactor it as you see fit?

Reply:
I can live with that, but please open an issue for the refactoring and assign it to me.

Reply:
Thanks, will do!